mne/beamformer/_lcmv.py

```diff
@@ -24,19 +24,47 @@
 from ..channels.channels import _contains_ch_type
 
 
-def _reg_pinv(x, reg):
-    """Compute a regularized pseudoinverse of a square array."""
-    if reg == 0:
-        covrank = estimate_rank(x, tol='auto', norm=False,
-                                return_singular=False)
-        if covrank < x.shape[0]:
-            warn('Covariance matrix is rank-deficient, but no regularization '
-                 'is done.')
+def _reg_pinv(x, reg, rcond=1e-15):
+    """Compute a regularized pseudoinverse of a square array.
+
+    Parameters
+    ----------
+    x : ndarray, shape (n, n)
+        Square array to invert.
+    reg : float
+        Regularization parameter.
+    rcond : float | 'auto'
+        Cutoff for small singular values. Singular values smaller (in modulus)
+        than `rcond` * largest_singular_value (again, in modulus) are set to
+        zero. Use 'auto' to attempt to automatically set a sane value. Defaults
+        to 1e-15.
+    """
+    covrank, s = estimate_rank(x, tol='auto', norm=False, return_singular=True)
 
-    # This adds it to the diagonal without using np.eye
+    # This adds the regularization without using np.eye
     d = reg * np.trace(x) / len(x)
+    x = x.copy()
     x.flat[::x.shape[0] + 1] += d
-    return linalg.pinv(x), d
+
+    if covrank < len(x):
+        if reg == 0:
+            warn('Covariance matrix is rank-deficient and no regularization '
+                 'is done.')
+
+        if rcond == 'auto':
+            # Reduce the toleration of the pseudo-inverse to force a solution
+            s = linalg.svd(x, compute_uv=False)
+            tol = s[covrank - 1:covrank + 1].mean()
+            tol = max(
+                tol,
+                len(x) * linalg.norm(x) * np.finfo(float).eps
+            )
+            rcond = tol / s.max()
```

Comment:
This computation seems non-standard, is there a good reason to diverge from what is usually done in `linalg` computations / has this been extensively tested?

Reply:
This is doing somethink akin to the `reduce_rank` parameter of the `make_lcmv` code. Beamformers like LCMV and DICS have a tendency to 'explode' and give nonsense results. The key point where this tends to happen is in the computation of the inverse. A run-of-the-mill `pinv` is in many cases not enough do deal with the rank-deficiency of the matrices.

I have the code like this, so it replicates Jan Kujala's pipeline, which is thoroughly tested.
There may be other (perhaps even better) ways of doing it, and in all likelihood we can use the same method for both LCMV and DICS beamformers, but this would need discussion and testing. So I would like to leave it for a follow-up PR.